Marketplace: Add no results page

raicem opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:
When we have no results as a result of a search term or a category selection, we show this message and the most popular products section. 

### How to test the changes in this Pull Request:

1.Build WooCommerce Admin on your local environment: `cd plugins/woocommerce-admin` then `nvm use && pnpm start`.
2. Visit WooCommerce > Marketplace > Extensions (eg. http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions&tab=extensions).
3. Search for some term (for example shipping or tax) and see the products listed on the page.
4. Search for some term that won't result in any results and see the new no results page.
5. Remove your search term and select the **Featured** category. **Featured** category, somehow doesn't have any results. See the error message now saying "We didn't find any results for "_featured". This is not a great message and I'm thinking about a possible solution to this.
    - I'm annoyed about the `_featured` bit in the message. The fact that it starts with an underscore makes it look bad to my eyes.

![image (8)](https://github.com/woocommerce/woocommerce/assets/10389957/53ed89aa-43b1-4417-8af0-09ade02359b8)

Dan-Q commented:
Not sure if this is just me, but I see the "no results" image at full-width. Tested in Firefox and Chrome; same thing -

![Uploading Screenshot 2023-08-14 at 14-05-21 Extensions ‹ WooCommerce ‹ woocommerce http __localhost 8888_.png…]()

Dan-Q commented:
As for the "_featured" message, we could do a hacky solution and remove leading underscores in the `category` parameter when displaying?

kdevnel commented on plugins/woocommerce-admin/client/marketplace/assets/images/no-results.svg:
I have a sneaky suspicion this image isn't unique to our use-case and is used elsewhere in WC Admin. It may even be a component we should pull in. I see something [used in WooCommerce Blocks](https://github.com/woocommerce/woocommerce-blocks/blob/4dd92c3e8d047dfead8135575b2d444c2ab3085d/assets/js/blocks/classic-template/product-search-results.ts#L51C23-L51C23) that might be similar, but I'm not certain and don't know how to find that block.

That said, without Poli around this week, I don't think it's worth holding off on merging what we have. It might just be worth checking when she's back so not a blocker

raicem replied:
I had a quick look in `woocommerce/woocommerce` and `woocommerce/woocommerce-blocks` repos but can't find it. I only searched for SVG files, using the GitHub's `language:svg` syntax.

kdevnel commented on plugins/woocommerce-admin/client/marketplace/components/extensions/extensions.tsx:
```diff
@@ -12,34 +12,48 @@ import CategorySelector from '../category-selector/category-selector';
 		title = sprintf(
 			// translators: %s: number of extensions
-			__( '%s extensions', 'woocommerce' ),
+			_n(
```
I ❤️  this translation method so much.

kdevnel reviewed (commented):
No blocking comments here. Everything looks good. 

Regarding the _featured, like Dan said, but is it possible to just strip **any** leading non-letter characters from the categories as you output them?

PS, don't forget to rebase on your updated parent branch so the new commits there don't show as part of this branch history.

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-list-content/no-results.tsx:
```diff
@@ -0,0 +1,110 @@
+				<img
+					className="woocommerce-marketplace__no-results__icon"
+					src={ NoResultsIcon }
+					alt="No results."
```
We could translate this.

Dan-Q replied:
```suggestion
					alt={ __( 'No results.', 'woocommerce' ) }
```

andfinally commented on plugins/woocommerce-admin/client/marketplace/components/product-list-content/no-results.tsx:
```diff
@@ -0,0 +1,110 @@
+		}
+
+		if ( ! productGroup ) {
+			return <></>;
```
Or just `</>`? I dunno if the linter will like that.

Dan-Q replied:
Don't think a self-closing `</>` is valid React (without an opening `<>`: I imagine it can't tell the difference between a closing empty tag and a self-closing empty tag). But you might be able to return `null`, or an empty string, or something?

andfinally replied:
Ah nice, fair enough. No biggie.

andfinally reviewed (approved):
Looks lovely! 👌 

We shouldn't include `_featured` in the categories we show! In the current marketplace, we use it to display the content we are now showing on the "Discover" tab.

<img width="500" src="https://github.com/woocommerce/woocommerce/assets/1647564/d3972daa-66f7-4a89-974f-28d745c79493">

I reckon it might mess up the history in the feature branch if we merge this without rebasing and squashing, but I realise time is short.

Dan-Q commented:
@andfinally said:
> We shouldn't include `_featured` in the categories we show! In the current marketplace, we use it to display the content we are now showing on the "Discover" tab.

We can fix this as follows, if everybody agrees - where we write:

```typescript
				const categories: Category[] = categoriesFromAPI.map(
					( categoryAPIItem: CategoryAPIItem ): Category => {
						return {
							...categoryAPIItem,
							selected: false,
						};
					}
				);
```

We could add a filter, like this:

```typescript
				const categories: Category[] = categoriesFromAPI.map(
					( categoryAPIItem: CategoryAPIItem ): Category => {
						return {
							...categoryAPIItem,
							selected: false,
						};
					}
				).filter(
					( category: Category ): boolean => {
						// The "featured" category is returned from the API for legacy reasons, but we don't need it:
						return category.slug != '_featured'
					}
				);
```

What do you think?

raicem committed:
Marketplace: Add no results page

raicem committed:
Marketplace: Add translation to the no results icon alt text

Co-authored-by: Dan Q <dan@danq.me>

raicem committed:
Marketplace: remove Featured category from category selector

It's not needed as a category option as the Discover page
shows the featured products. So we filter it out.

raicem commented:
Thank you all for your feedback and suggestions. I think I went through all. The only thing I couldn't figure out is the giant icon issue reported by @Dan-Q. I can't replicate it so far, in various browsers. Also, Gareth and Kyle also didn't report anything similar. Could it be a caching issue I wonder?

Dan-Q commented:
> The only thing I couldn't figure out is the giant icon issue reported by @Dan-Q.

I rebuilt my dev env and it went away. 🤷 Must've been _something_ at my end, though I don't know what!

Dan-Q reviewed (approved):
Looking great!

kdevnel reviewed (approved):
Looks good!
